java/client/src/test/java/glide/ExceptionHandlingTests.java

```diff
@@ -0,0 +1,273 @@
+        assertFalse(channelHandler.wasClosed);
+    }
+
+    @Test
+    @SneakyThrows
+    public void connection_manager_rethrows_non_RedisException_too() {
+        var callbackDispatcher = new TestCallbackDispatcher(new IOException("TEST"));
+        var channelHandler = new TestChannelHandler(callbackDispatcher);
+        var connectionManager = new ConnectionManager(channelHandler);
+
+        var future = connectionManager.connectToRedis(createDummyConfig());
+        callbackDispatcher.completeRequest(null);
+
+        var exception = assertThrows(ExecutionException.class, future::get);
+        // a IOException thrown from CallbackDispatcher::completeRequest and then wrapped
+        // by a RuntimeException and rethrown by ConnectionManager::exceptionHandler
+        assertTrue(exception.getCause() instanceof RuntimeException);
+        assertTrue(exception.getCause().getCause() instanceof IOException);
+        // check the channel
+        assertTrue(channelHandler.wasClosed);
+    }
+
+    @Test
+    @SneakyThrows
+    public void callback_dispatcher_handles_response_with_closing_error() {
+        var callbackDispatcher = new CallbackDispatcher();
+        var channelHandler = new TestChannelHandler(callbackDispatcher);
+        var connectionManager = new ConnectionManager(channelHandler);
+
+        var future1 = connectionManager.connectToRedis(createDummyConfig());
+        var future2 = connectionManager.connectToRedis(createDummyConfig());
+        var response = Response.newBuilder().setCallbackIdx(42).setClosingError("TEST").build();
+        callbackDispatcher.completeRequest(response);
+
+        var exception = assertThrows(ExecutionException.class, future1::get);
+        // a ClosingException thrown from CallbackDispatcher::completeRequest and then
+        // rethrown by CommandManager::exceptionHandler
+        assertTrue(exception.getCause() instanceof ClosingException);
+        // check the channel
+        assertTrue(channelHandler.wasClosed);
+
+        // all pending requests should be aborted once ClosingError received in callback dispatcher
+        exception = assertThrows(ExecutionException.class, future2::get);
+        // or could be cancelled in CallbackDispatcher::shutdownGracefully
+        // cancellation overwrites previous status, so we may not get ClosingException due to a race
+        assertTrue(
+                exception.getCause() instanceof ClosingException
+                        || exception.getCause() instanceof CancellationException);
+    }
+
+    @Test
+    @SneakyThrows
+    public void callback_dispatcher_handles_response_with_request_error() {
+        var callbackDispatcher = new CallbackDispatcher();
+        var channelHandler = new TestChannelHandler(callbackDispatcher);
+        var commandManager = new CommandManager(channelHandler);
+
+        var protobufErrorsToJavaClientErrors =
+                Map.of(
+                        Unspecified, RequestException.class,
```

Comment:
this list is nice, but what mechanism ensures that if there's a new exception type the list will update? Can't we iterate over the types using reflection?

Reply:
We can iterate with reflection, but mapping is still hardcoded. I'll try to make mapping as a class property to avoid duplicating it in tests.

Reply:
Reflection is ugly and requires additional error handling. For nice mapping we should update the enum, but it is managed by protobuf.
I keep this unchanged.